debezium-core/src/main/java/io/debezium/connector/base/ChangeEventQueue.java

```diff
@@ -227,31 +232,55 @@ protected void doEnqueue(T record) throws InterruptedException {
 
         try {
             LOGGER.debug("polling records...");
-            List<T> records = new ArrayList<>();
             final Timer timeout = Threads.timer(Clock.SYSTEM, Temporals.min(pollInterval, ConfigurationDefaults.RETURN_CONTROL_INTERVAL));
-            while (!timeout.expired() && queue.drainTo(records, maxBatchSize) == 0) {
-                throwProducerExceptionIfPresent();
-
-                LOGGER.debug("no records available yet, sleeping a bit...");
-                // no records yet, so wait a bit
-                metronome.pause();
-                LOGGER.debug("checking for more records...");
-            }
-            if (maxQueueSizeInBytes > 0 && records.size() > 0) {
-                records.parallelStream().forEach((record) -> {
-                    if (objectMap.containsKey(record)) {
-                        currentQueueSizeInBytes.addAndGet(-objectMap.get(record));
-                        objectMap.remove(record);
+            synchronized (this) {
+                List<T> records = new ArrayList<>(Math.min(maxBatchSize, queue.size()));
```

Comment:
@krnaveen14 The PR looks to be really tgought about in detail. It is quite understandable and I am going to merge it to get it inculde in tomorrow release.
Still there is one question for you. Should not this line alway allocat the `maxBatchSize` as the initial capacity? As you are striving to build the full batch it can happen that the backing array would need to be reallocated mutliple times while the buffer will be growing. Is not it so?